internal/storage/prefixdb.go

```diff
@@ -0,0 +1,632 @@
+package storage
+
+import (
+	"bytes"
+	"encoding/hex"
+	"fmt"
+	"sync"
+)
+
+// PrefixDB provides a logical database by wrapping a namespace of another database.
+// It allows for the creation of multiple logical databases on top of a single
+// physical database by automatically prefixing keys with a specified namespace.
+//
+// It wraps a [DB] interface, which represents the underlying database, and
+// delegates operations to the wrapped database after prepending the prefix to the
+// key. That is, PrefixDB ensures that all keys written to or read from the
+// underlying database are scoped to the provided prefix.
+//
+// Concurrent access to a PrefixDB is safe as long as the underlying [DB]
+// implementation supports concurrent operations.
+//
+// Example usage:
+//
+//	db := NewPrefixDB(baseDB, []byte("namespace:"))
+//	err := db.Set([]byte("key"), []byte("value"))
+//	if err != nil {
+//		// handle error
+//	}
+//
+// In this example, the key "key" will be stored in 'baseDB' with the actual key
+// being "namespace:key".
+type PrefixDB struct {
+	mu     sync.Mutex
+	db     DB
+	prefix []byte
+}
+
+// compile-time check: does *PrefixDB satisfy the DB interface?
+var _ DB = (*PrefixDB)(nil)
+
+// NewPrefixDB returns a new PrefixDB wrapping the given database and scoping it to
+// the given prefix.
+func NewPrefixDB(db DB, prefix []byte) *PrefixDB {
+	return &PrefixDB{
+		prefix: prefix,
```

Comment:
Should we check whether the given prefix is empty, or is it a valid case? I don't think there is much point to a `PrefixDB` without a prefix 😅

Reply:
We definitely should